textworld/utils.py

```diff
@@ -192,6 +192,39 @@ def uniquify(seq):
     return [x for x in seq if x not in seen and not seen.add(x)]
 
 
+def _unique_product_recursive(pools, result, i):
+    if i >= len(pools):
+        yield tuple(result)
+        return
+
+    for e in pools[i]:
+        if e not in result:
+            result[i] = e
+            yield from _unique_product_recursive(pools, result, i + 1)
+            result[i] = None
+
+
+def unique_product(*iterables):
+    """ Cartesian product of input iterables with pruning.
+
+    This method prunes any product tuple with duplicate elements in it.
+
+    Example:
+        unique_product('ABC', 'Ax', 'xy') --> Axy BAx BAy Bxy CAx CAy Cxy
+
+    Notes:
+        This method is faster than the following equivalent code:
+
+        >>> for result in itertools.product(*args):
+        >>>     if len(set(result)) == len(result):
+        >>>         yield result
+
+    """
+    pools = [tuple(pool) for pool in iterables]
+    result = [None] * len(pools)
```

Comment:
This will cause issues if `None` appears in the input.  It needs to use a sentinel that isn't equal to anything else, maybe something like `_SENTINEL = object()`.  Then use `_SENTINEL` instead of `None` here and on line 204 (`result[i] = None`)